### PR TITLE
EVAKA-4012: improve and unify planned occupancy logic

### PR DESCRIPTION
**Pull request by Joosakur**
#### Summary
Changed planned occupancy logic (now uses 'prepare for the worst case scenario' principle) and unified it between unit view and report:

- If child has overlapping placement and placement_plan to different units, child counts into both units
- If child has overlapping placement and placement_plan to same unit, the one with higher capacity need wins

#### Checklist for pull request creator

- [ ] The code is consistent with the existing code base
- [ ] Tests have been written for the change (e2e, integration tests)
- [ ] The change has been tested locally
- [ ] The change conforms to the UX specifications
- [ ] The code is self-documenting or has been documented sufficiently, e.g., in the README
- [ ] The branch has been rebased against master before the PR was created

#### Checklist for pull request reviewer (copy to review text box)

```
- [ ] All changes in all changed files have been reviewed
- [ ] The code is consistent with the existing code base
- [ ] Tests have been written for the change (e2e, integration tests)
- [ ] The change has been tested locally
- [ ] The change conforms to the UX specifications
- [ ] The code is self-documenting or has been documented sufficiently, e.g., in the README
- [ ] The PR branch has been rebased against master and force pushed if necessary before merging
```

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/occupancy/Occupancy.kt [hunk 1]
```diff
@@ -225,13 +225,17 @@ private val plannedMaxCoefficients =
...
-        coalesce(p.child_id, c.child_id) AS child_id,
-        coalesce(p.coefficient, c.coefficient) AS coefficient
-    FROM coefficients c
-    FULL JOIN planned_max_coefficients p ON c.day = p.day AND c.child_id = p.child_id
+    SELECT unit_id, day, child_id, max(coefficient) AS coefficient
+    FROM (
+        SELECT unit_id, day, child_id, coefficient
+        FROM planned_max_coefficients
+
+        UNION DISTINCT
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/occupancy/Occupancy.kt:233]: If you're getting the max coefficient for the day anyway, is the `DISTINCT` useful?

Comment (Joosakur) [reply service/src/main/kotlin/fi/espoo/evaka/occupancy/Occupancy.kt:233]: hmm, yeah not anymore, changing to UNION ALL

File: service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt [hunk 2]
```diff
@@ -326,6 +422,83 @@ class PlannedOccupancyTest : FullApplicationTest() {
...
+        assertEquals(1.0, reportResult1?.sum)
+        assertEquals(1, reportResult1?.headcount)
+
+        val reportResult2 = fetchAndParsePlannedOccupancyReportForUnitDay(testAreaId, daycareId2, period2.start)
+        assertEquals(0.54, reportResult2?.sum)
+        assertEquals(1, reportResult2?.headcount)
+    }
+
+    @Test
+    fun `when child is planned to transfer to same unit the higher capacity wins`() {
```
Comment (jaahon) [inline service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:476]: I think a test where the situation is opposite (planned placement has higher occupancy) would be valuable as well.

**Review by jaahon — COMMENTED** [review]
By the way, why don't you use the `createPlanOccupancyTestFixture` function for data initialization?

File: service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt (hunk 2 above)
Comment (Joosakur) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:476]: hmm actually the tests `occupancy calculation is correct for child over 3 years old with full time placement plan over an existing part time placement`  and `occupancy calculation is correct for child over 3 years old with part time placement plan over an existing full time placement`  already cover both ways so this test can be deleted as duplicate?

Comment (jaahon) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:476]: I guess so, but the assertion in `occupancy calculation is correct for child over 3 years old with part time placement plan over an existing full time placement` expects the opposite of what you intend (I think) 😄

Comment (Joosakur) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:476]: it used to, but I changed the assertion (line 351)

Comment (jaahon) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:476]: Oh yeah, sorry, I was looking at master.

File: service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt [hunk 3]
```diff
@@ -37,7 +48,9 @@ class PlannedOccupancyTest : FullApplicationTest() {
         jdbi.handle(::resetDatabase)
     }
 
-    private val defaultPeriod = ClosedPeriod(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 12, 31))
+    private val defaultPeriod = ClosedPeriod(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 1, 31))
+    val period1 = ClosedPeriod(defaultPeriod.start, LocalDate.of(2019, 1, 15))
+    val period2 = ClosedPeriod(LocalDate.of(2019, 1, 16), defaultPeriod.end)
```
Comment (jaahon) [inline service/src/integrationTest/kotlin/fi/espoo/evaka/occupancy/PlannedOccupancyTest.kt:53]: This is very minor, but should these be named something like `defaultPeriodSplit1` and `defaultPeriodSplit2` so it's immediately apparent that they are two halves of `defaultPeriod`?

**Review by jaahon — APPROVED** [review]
